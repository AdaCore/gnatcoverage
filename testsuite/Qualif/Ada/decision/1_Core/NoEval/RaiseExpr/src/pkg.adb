pragma Ada_2022;

package body Pkg is

   function To_Pos (X : Integer; Do_Raise : Boolean) return Integer is
   begin
      if (Do_Raise and then raise Program_Error) or else X >= 0 then -- # eval
         return X;                                                   -- # true
      else
         return -X;                                                  -- # false
      end if;
   end To_Pos;

end Pkg;
