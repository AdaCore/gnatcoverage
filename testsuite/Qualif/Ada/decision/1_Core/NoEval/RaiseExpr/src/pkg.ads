package Pkg is

   function To_Pos (X : Integer; Do_Raise : Boolean) return Integer;
   --  Return the absolute value of X. If Do_Raise is True, and exception is
   --  raised in the main decision.

end Pkg;
