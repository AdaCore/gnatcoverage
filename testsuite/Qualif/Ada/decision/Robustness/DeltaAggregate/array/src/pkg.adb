pragma Ada_2022;

package body Pkg is

   --------------------------
   -- Copy_With_Abs_Update --
   --------------------------

   function Copy_With_Abs_Update
     (Input     : Composite;
      Item      : Item_Type;
      Use_Cache : Boolean := False) return Composite
   is
      Elem : integer renames -- # st
        Input (if Item = First then 1 else 2);  -- # comp_expr
   begin
      return  -- # st
        [(if Use_Cache then Cached_Comp else Input) with delta  -- # base_expr_first
         (if Item = First then 1 else 2) =>                     -- # comp_expr
         (if Elem >= 0 then Elem else -Elem)];                  -- # value_expr_first
   end Copy_With_Abs_Update;
end Pkg;
