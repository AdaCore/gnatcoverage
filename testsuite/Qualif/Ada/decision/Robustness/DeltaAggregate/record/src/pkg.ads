pragma Ada_2012;

package Pkg is

   type Item_Type is (First, Last);

   type Composite is private;

   Cached_Comp : constant Composite;

   function Make (First, Last : Integer) return Composite;

   function Eq (Comp : Composite; First, Last : Integer) return Boolean;

   function Copy_With_Abs_Update
     (Input     : Composite;
      Item      : Item_Type;
      Use_Cache : Boolean := False) return Composite;
   --  Return a copy of Input or Cached_Comp depending on the value of
   --  Use_Cache, with either the "First" or "Last" component being updated to
   --  its absolute value.
private
   type Composite is record
      First : Integer;
      Last  : Integer;
   end record;

   Cached_Comp : constant Composite := (0, 0);

   ----------
   -- Make --
   ----------

   function Make (First, Last : Integer) return Composite is ((First, Last));

   --------
   -- Eq --
   --------

   function Eq (Comp : Composite; First, Last : Integer) return Boolean is
     (Comp.First = First and then Comp.Last = Last);

end Pkg;
