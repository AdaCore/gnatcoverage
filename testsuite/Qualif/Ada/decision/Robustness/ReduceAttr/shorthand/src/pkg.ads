package Pkg is

   type Arr_T is array (Positive range <>) of Integer;

   type Op_T is (Sum, Max);

   function Compute (Inp : Arr_T; Op : Op_T) return Integer;
   --  Compute either the sum or the maximum of the Element in Inp based on the
   --  value of Op.

end Pkg;
