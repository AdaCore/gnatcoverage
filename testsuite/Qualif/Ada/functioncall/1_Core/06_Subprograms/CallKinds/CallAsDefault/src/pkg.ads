pragma Ada_2012;

package Pkg is

   function Forty_Two return Integer is (42);                       -- # decl

   --  Call as a record component's default value
   type Rec is record
      Number : Integer := Forty_Two;                                -- # comp
    end record;

   --  Call as a subprogram parameter's default value
   function Add (R : Rec; I : Integer := Forty_Two) return Integer  -- # spec
   is (R.Number + I);                                               -- # body

 end Pkg;
