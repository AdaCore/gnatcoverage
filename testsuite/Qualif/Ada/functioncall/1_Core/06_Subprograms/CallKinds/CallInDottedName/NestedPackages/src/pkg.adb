pragma Ada_2012;

package body Pkg is
    package body A is
        package body B is
            procedure P (X : out Integer) is            -- # v_fun
            begin
                X := 42;                                -- # v_stmt
            end P;

            function P (X : Integer) return Integer is  -- # v_fun
            begin
                return 42;                              -- # v_stmt
            end P;
        end B;
    end A;
end Pkg;
