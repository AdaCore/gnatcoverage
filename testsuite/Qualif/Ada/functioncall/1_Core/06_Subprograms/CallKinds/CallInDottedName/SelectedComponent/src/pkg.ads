pragma Ada_2012;

package Pkg is

   type Point is record                                       -- # decl
      X, Y : Float;                                           -- # decl
   end record;                                                -- # decl

   function Set (X, Y : Float) return Point is (X, Y);        -- # set

end Pkg;
