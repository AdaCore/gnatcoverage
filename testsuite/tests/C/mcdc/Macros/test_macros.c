#include "pkg.h"
#include <stdio.h>

#define PRINT_HW printf ("Hello world!\n");
#define PRINT printf
#define HW "Hello world!\n"
#define PRINT_HW_PRINT_HW                                                     \
  PRINT_HW;                                                                   \
  PRINT_HW
#define WRAPPER_PRINT_HW PRINT_HW
#define ID(x) x
#define AND(x, y) ID (x) && y
#define WRAP_AND(x, y) AND (x, y)
#define WRAP_CMDLINE_MACRO_STMT CMDLINE_MACRO_STMT

void
macro_stmts ()
{
  PRINT_HW_PKG;
  PRINT_HW;
  WRAPPER_PRINT_HW;
  PRINT ("Hello world!\n");
  PRINT (HW);
  PRINT_HW_PRINT_HW;
  WRAP_CMDLINE_MACRO_STMT;
  CMDLINE_MACRO_STMT;
  CMDLINE_MACRO_DECL (a);
  CMDLINE_MACRO_NO_VALUE (int b =);
  // Test that coverage obligations coming from built-in macros are reported
  // correctly.
  __linux + 1;
}

void
macro_decision_violations (int a, int b)
{
  if (WRAP_AND (ID (a), b))
    PRINT_HW;
}

int
main ()
{
  macro_decision_violations (1, 0);
  macro_decision_violations (1, 1);
  return 0;
}
