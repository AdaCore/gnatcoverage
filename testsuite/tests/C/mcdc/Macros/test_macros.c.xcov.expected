25% of 16 lines covered
29% statement coverage (5 out of 17)
100% decision coverage (1 out of 1)
50% MC/DC coverage (1 out of 2)

Coverage level: stmt+mcdc
   1 .: #include "pkg.h"
   2 .: #include <stdio.h>
   3 .: 
   4 .: #define PRINT_HW printf ("Hello world!\n");
   5 .: #define PRINT printf
   6 .: #define HW "Hello world!\n"
   7 .: #define PRINT_HW_PRINT_HW                                                     \
   8 .:   PRINT_HW;                                                                   \
   9 .:   PRINT_HW
  10 .: #define WRAPPER_PRINT_HW PRINT_HW
  11 .: #define ID(x) x
  12 .: #define AND(x, y) ID (x) && y
  13 .: #define WRAP_AND(x, y) AND (x, y)
  14 .: #define WRAP_CMDLINE_MACRO_STMT CMDLINE_MACRO_STMT
  15 .: 
  16 .: void
  17 .: macro_stmts ()
  18 .: {
  19 -:   PRINT_HW_PKG;
statement "printf ("..." at 19:3 not executed
  note: in definition of macro PRINT_HW_PKG at location pkg.h:1:22
  note: from expansion of macro PRINT_HW_PKG at location test_macros.c:19:3
  20 -:   PRINT_HW;
statement "printf ("..." at 20:3 not executed
  note: in definition of macro PRINT_HW at location test_macros.c:4:18
  note: from expansion of macro PRINT_HW at location test_macros.c:20:3
  21 -:   WRAPPER_PRINT_HW;
statement "printf ("..." at 21:3 not executed
  note: in definition of macro PRINT_HW at location test_macros.c:4:18
  note: from expansion of macro PRINT_HW at location test_macros.c:10:26
  note: from expansion of macro WRAPPER_PRINT_HW at location test_macros.c:21:3
  22 -:   PRINT ("Hello world!\n");
statement "printf ("..." at 22:3 not executed
  note: in definition of macro PRINT at location test_macros.c:5:15
  note: from expansion of macro PRINT at location test_macros.c:22:3
  23 -:   PRINT (HW);
statement "printf ("..." at 23:3 not executed
  note: in definition of macro PRINT at location test_macros.c:5:15
  note: from expansion of macro PRINT at location test_macros.c:23:3
  24 -:   PRINT_HW_PRINT_HW;
statement "printf ("..." at 24:3 not executed
  note: in definition of macro PRINT_HW at location test_macros.c:4:18
  note: from expansion of macro PRINT_HW at location test_macros.c:8:3
  note: from expansion of macro PRINT_HW_PRINT_HW at location test_macros.c:24:3
statement "printf ("..." at 24:3 not executed
  note: in definition of macro PRINT_HW at location test_macros.c:4:18
  note: from expansion of macro PRINT_HW at location test_macros.c:9:3
  note: from expansion of macro PRINT_HW_PRINT_HW at location test_macros.c:24:3
  25 -:   WRAP_CMDLINE_MACRO_STMT;
statement "printf ("..." at 25:3 not executed
  note: in definition of macro CMDLINE_MACRO_STMT at location <command line>
  note: from expansion of macro CMDLINE_MACRO_STMT at location test_macros.c:14:33
  note: from expansion of macro WRAP_CMDLINE_MACRO_STMT at location test_macros.c:25:3
  26 -:   CMDLINE_MACRO_STMT;
statement "printf ("..." at 26:3 not executed
  note: in definition of macro CMDLINE_MACRO_STMT at location <command line>
  note: from expansion of macro CMDLINE_MACRO_STMT at location test_macros.c:26:3
  27 -:   CMDLINE_MACRO_DECL (a);
statement "int a;" at 27:3 not executed
  note: in definition of macro CMDLINE_MACRO_DECL at location <command line>
  note: from expansion of macro CMDLINE_MACRO_DECL at location test_macros.c:27:3
  28 -:   CMDLINE_MACRO_NO_VALUE (int b =);
statement "int b = 1..." at 28:3 not executed
  note: in definition of macro CMDLINE_MACRO_NO_VALUE at location <command line>
  note: from expansion of macro CMDLINE_MACRO_NO_VALUE at location test_macros.c:28:3
  29 .:   // Test that coverage obligations coming from built-in macros are reported
  30 .:   // correctly.
  31 -:   __linux + 1;
statement "1 + 1" at 31:3 not executed
  note: in definition of macro __linux at location <built-in>
  note: from expansion of macro __linux at location test_macros.c:31:3
  32 .: }
  33 .: 
  34 .: void
  35 .: macro_decision_violations (int a, int b)
  36 .: {
  37 !:   if (WRAP_AND (ID (a), b))
condition "a" at 37:7 has no independent influence pair, MC/DC not achieved
  note: in definition of macro ID at location test_macros.c:11:15
  note: from expansion of macro ID at location test_macros.c:12:19
  note: from expansion of macro AND at location test_macros.c:13:24
  note: from expansion of macro WRAP_AND at location test_macros.c:37:7
  note: from expansion of macro ID at location test_macros.c:37:17
  38 +:     PRINT_HW;
  39 .: }
  40 .: 
  41 .: int
  42 .: main ()
  43 .: {
  44 +:   macro_decision_violations (1, 0);
  45 +:   macro_decision_violations (1, 1);
  46 +:   return 0;
  47 .: }
